### PR TITLE
Performance optimization creation of Event Engine Genesis #626

### LINKED ISSUE #626: Performance optimization for create event engine genesis

### PR DESCRIPTION
**Pull request by s-medvedev**
Resolves #626
Decrease time of execution `create-genesis-ee` from 45 minutes to 15.
Remove `count` argument from `start_section` (write actual rows count in `finish_section`).

### CODE REVIEW

File: programs/create-genesis-ee/genesis_ee_builder.hpp [hunk 1]
```diff
@@ -36,8 +36,8 @@ class genesis_ee_builder final {
     void process_delete_reblogs();
     void process_follows();
 
-    variants build_votes(uint64_t msg_hash, operation_number msg_created);
-    variants build_reblogs(uint64_t msg_hash, operation_number msg_created, bfs::fstream& dump_reblogs);
+    void build_votes(std::vector<vote_info>& votes, uint64_t msg_hash, operation_number msg_created);
+    void build_reblogs(std::vector<reblog_info>& reblogsi, uint64_t msg_hash, operation_number msg_created, bfs::ifstream& dump_reblogs);
```
Comment (afalaleev) [inline programs/create-genesis-ee/genesis_ee_builder.hpp]: reblogsi

File: libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp [hunk 2]
```diff
@@ -40,36 +40,43 @@ struct ee_genesis_serializer {
...
-            auto pos = out.tellp();
-            out.seekp(_section_offset);
-            fc::raw::pack(out, _section);
-            out.seekp(pos);
+        auto pos = out.tellp();
+        out.seekp(_section_offset);
 
-            wlog("Finished with count: ${count}", ("count", _section.count));
-        }
+        _section.count += _count;
```
Comment (boostmultifruit) [inline libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp]: `_count` field is useless now. Just increment `_section.count`.

File: libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp [hunk 3]
```diff
@@ -40,36 +40,43 @@ struct ee_genesis_serializer {
...
+        fc::raw::pack(out, _section);
 
-        EOS_ASSERT(_count == _section.count, ee_genesis_exception, "Section contains wrong number of rows",
-            ("section",_section)("diff", _section.count - _count));
```
Comment (boostmultifruit) [inline libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp]: This check is not so useful, but we can keep such feature by adding optional argument `check_count` to `finish_section`.

File: plugins/event_engine_plugin/event_engine_plugin.cpp [hunk 4]
```diff
@@ -214,37 +215,31 @@ void event_engine_plugin_impl::send_genesis_file(const bfs::path& genesis_file)
...
 
     std::cout << "Reading event engine genesis data from " << genesis_file << "..." << std::endl;
-    bfs::ifstream in(genesis_file);
-    ee_genesis_header h{"", 0};
+    boost::iostreams::mapped_file mfile;
+    mfile.open(genesis_file, boost::iostreams::mapped_file::readonly);
+    EOS_ASSERT(mfile.is_open(), ee_extract_genesis_exception, "File not opened");
 
-    in.read((char*)&h, sizeof(h));
+    ee_genesis_header &h = *(ee_genesis_header*)mfile.const_data();
```
Comment (boostmultifruit) [inline plugins/event_engine_plugin/event_engine_plugin.cpp]: `const auto`?

Comment (s-medvedev) [reply plugins/event_engine_plugin/event_engine_plugin.cpp]: `const` - Ok. but in this place I need guarantees that this is exactly this type, so without `auto`.

File: libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp (hunk 2 above)
Comment (s-medvedev) [reply libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp]: Fixed

File: programs/create-genesis-ee/genesis_ee_builder.hpp (hunk 1 above)
Comment (s-medvedev) [reply programs/create-genesis-ee/genesis_ee_builder.hpp]: Fixed

File: plugins/event_engine_plugin/event_engine_plugin.cpp (hunk 4 above)
Comment (boostmultifruit) [reply plugins/event_engine_plugin/event_engine_plugin.cpp]: @s-medvedev C++ is not interpreting :) There will be `ee_genesis_header` anyway.

File: libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp [hunk 5]
```diff
@@ -40,37 +39,49 @@ struct ee_genesis_serializer {
...
+        auto pos = out.tellp();
+        out.seekp(_section_offset);
+
+        fc::raw::pack(out, _section);
+
+        out.seekp(pos);
+        wlog("Finished with count: ${count}", ("count", _section.count));
```
Comment (boostmultifruit) [inline libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp]: Output is:
```
Reading operation dump from "/root/1290/golos/build2/programs/golosd/witness_node_data_dir/operation_dump"...
-> Reading comments...
-> Reading comment deletions...
-> Reading rewards...
-> Reading votes...
-> Reading reblogs...
-> Reading delete reblogs...
-> Reading follows...
Writing genesis to "/root/cyberway/events-genesis"...
-> Writing messages...
warn  2019-05-08T09:01:02.185 thread-0  ee_genesis_serializer.:46     start_section        ] Starting section: {"code":"gls.publish","name":"message","abi_type":"message_info","count":0}
-> Writing transfers...
warn  2019-05-08T09:01:02.298 thread-0  ee_genesis_serializer.:46     start_section        ] Starting section: {"code":"cyber.token","name":"transfer","abi_type":"transfer","count":0}
-> Writing pinblocks...
warn  2019-05-08T09:01:02.298 thread-0  ee_genesis_serializer.:46     start_section        ] Starting section: {"code":"gls.social","name":"pin","abi_type":"pin","count":0}
warn  2019-05-08T09:01:02.299 thread-0  ee_genesis_serializer.:69     finish_section       ] Finished with count: 4442
warn  2019-05-08T09:01:02.299 thread-0  ee_genesis_serializer.:46     start_section        ] Starting section: {"code":"gls.social","name":"block","abi_type":"block","count":0}
warn  2019-05-08T09:01:02.300 thread-0  ee_genesis_serializer.:69     finish_section       ] Finished with count: 7126
warn  2019-05-08T09:01:02.300 thread-0  ee_genesis_serializer.:69     finish_section       ] Finished with count: 20
warn  2019-05-08T09:01:02.300 thread-0  ee_genesis_serializer.:69     finish_section       ] Finished with count: 32
```

Comment (s-medvedev) [reply libraries/chain/include/cyberway/genesis/ee_genesis_serializer.hpp]: Fixed

**Pull request merged**